### PR TITLE
An attempt has been made to fix species related shitcode

### PR DESCRIPTION
**Pull request by eeSPee**
## About The Pull Request

Trait Stomach_Brain, indicates the species/creature has the stomach in their brain.
Organ zone is now dynamic. This caused bugs with IPC brains who were inserted in chest. 
Species Organs also dynamic, because the code was a clusterfuck.
General Fixes

## Why It's Good For The Game

Should fix shitcode and IPC related bugs.

## Changelog
:cl:
fix: some IPC related bugs
fix: code in species should be smoother
add: IPCs take stam damage when they lack a stomach
/:cl:

### CODE REVIEW

File: code/datums/diseases/gastrolisis.dm [hunk 1]
```diff
@@ -72,11 +72,13 @@
...
 		if(tongue)
-			var/obj/item/organ/tongue/new_tongue = new H.dna.species.mutanttongue ()
+			var/o_type = H.dna.species.get_species_organs() [ORGAN_SLOT_TONGUE]
+			var/obj/item/organ/tongue/new_tongue = new o_type()
 			new_tongue.Insert(H)
 		var/obj/item/organ/eyes/eyes = locate(/obj/item/organ/eyes/snail) in H.internal_organs
 		if(eyes)
-			var/obj/item/organ/eyes/new_eyes = new H.dna.species.mutanteyes ()
+			var/o_type = H.dna.species.get_species_organs() [ORGAN_SLOT_EYES]
+			var/obj/item/organ/eyes/new_eyes = new o_type ()
```
Comment (PowerfulBacon) [inline code/datums/diseases/gastrolisis.dm:81]: This code was better before.

File: code/game/objects/items/devices/scanners.dm [hunk 2]
```diff
@@ -339,26 +339,12 @@ GENE SCANNER
...
-		else if (S.mutantliver != initial(S.mutantliver))
-			mutant = TRUE
-		else if (S.mutantstomach != initial(S.mutantstomach))
-			mutant = TRUE
+		else
+			var/list/base_organs = S.get_species_organs()
+			for (var/organ_type in base_organs)
+				if (!H.getorgan(organ_type))
+					mutant = TRUE
+					continue
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/scanners.dm:347]:
```suggestion
					break
```

File: code/modules/antagonists/changeling/powers/regenerate.dm [hunk 3]
```diff
@@ -26,8 +26,9 @@
 			C.regenerate_limbs(1)
 		if(!user.getorganslot(ORGAN_SLOT_BRAIN))
 			var/obj/item/organ/brain/B
-			if(C.has_dna() && C.dna.species.mutant_brain)
-				B = new C.dna.species.mutant_brain()
+			if(C.has_dna() && C.dna.species.get_species_organs()[ORGAN_SLOT_BRAIN])
+				var/organ = C.dna.species.get_species_organs()[ORGAN_SLOT_BRAIN]
+				B = new organ()
```
Comment (PowerfulBacon) [inline code/modules/antagonists/changeling/powers/regenerate.dm:31]: Code was better before, checking 1 variable > running a proc that checks a bunch of variables, assigning them to a list, then looking in the lost for the 1 variable you need.

File: code/modules/mob/living/carbon/human/species.dm [hunk 4]
```diff
@@ -143,130 +130,55 @@ GLOBAL_LIST_EMPTY(roundstart_races)
...
-			var/obj/item/organ/I = C.getorgan(mutantorgan)
-			if(I)
-				I.Remove(C)
-				QDEL_NULL(I)
-
-	for(var/path in mutant_organs)
-		var/obj/item/organ/I = new path()
-		I.Insert(C)
+	var/list/species_organs = get_species_organs()
+	for (var/ORGAN_SLOT in list(ORGAN_SLOT_HEART,ORGAN_SLOT_LUNGS,ORGAN_SLOT_APPENDIX,ORGAN_SLOT_LIVER,ORGAN_SLOT_STOMACH,ORGAN_SLOT_TAIL,ORGAN_SLOT_WINGS,ORGAN_SLOT_ADAMANTINE_RESONATOR,ORGAN_SLOT_RIGHT_ARM_AUG))
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/human/species.dm:150]: This list should be a well named define, to make it easier if adding new organ slots in future

Comment (PowerfulBacon) [reply code/modules/mob/living/carbon/human/species.dm:150]: Also the variable should not be in full caps

File: code/modules/mob/living/carbon/human/species.dm [hunk 5]
```diff
@@ -143,130 +130,55 @@ GLOBAL_LIST_EMPTY(roundstart_races)
...
+				var/obj/item/organ/new_O = new organt()
+				new_O.Insert(C,TRUE,FALSE)
+			else 
+				var/obj/item/organ/old_O = C.getorganslot(ORGAN_SLOT)
+				if (old_O)
+					old_O.Remove(C,TRUE)
+					QDEL_NULL(old_O)
+			
+	if(C.get_bodypart(BODY_ZONE_HEAD))	
+		for (var/ORGAN_SLOT in list(ORGAN_SLOT_EYES,ORGAN_SLOT_EARS,ORGAN_SLOT_TONGUE,ORGAN_SLOT_BRAIN,ORGAN_SLOT_VOICE,ORGAN_SLOT_BRAIN))
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/human/species.dm:167]: Same here

File: code/modules/mob/living/carbon/human/species_types/abductors.dm [hunk 6]
```diff
@@ -17,3 +16,8 @@
 	. = ..()
 	var/datum/atom_hud/abductor_hud = GLOB.huds[DATA_HUD_ABDUCTOR]
 	abductor_hud.remove_hud_from(C)
+
+/datum/species/abductor/get_species_organs()
+	var/list/organs = ..()
+	organs[ORGAN_SLOT_TONGUE] = /obj/item/organ/tongue/abductor
+	return organs
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/human/species_types/abductors.dm:23]: Again, code was better before as single variable checks are better than dumping it all into a list

File: code/modules/mob/living/carbon/human/species_types/android.dm [hunk 7]
```diff
@@ -28,3 +27,8 @@
 	for(var/X in C.bodyparts)
 		var/obj/item/bodypart/O = X
 		O.change_bodypart_status(BODYPART_ORGANIC,FALSE, TRUE)
+
+/datum/species/android/get_species_organs()
+	var/list/organs = ..()
+	organs[ORGAN_SLOT_TONGUE] = /obj/item/organ/tongue/robot
+	return organs
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/human/species_types/android.dm:34]: see above

File: code/modules/mob/living/carbon/human/species_types/apid.dm [hunk 8]
```diff
@@ -43,4 +39,12 @@
...
\ No newline at end of file
+		H.reagents.remove_reagent(chem.type, REAGENTS_METABOLISM)
+		
+/datum/species/apid/get_species_organs()
+	var/list/organs = ..()
+	organs[ORGAN_SLOT_LUNGS] = /obj/item/organ/lungs/apid
+	organs[ORGAN_SLOT_EYES] = /obj/item/organ/eyes/apid
+	organs[ORGAN_SLOT_WINGS] = /obj/item/organ/wings/bee
+	organs[ORGAN_SLOT_TONGUE] = /obj/item/organ/tongue/bee
+	return organs
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/human/species_types/apid.dm:50]: See above

File: code/modules/mob/living/carbon/human/species_types/dullahan.dm [hunk 9]
```diff
@@ -139,3 +135,11 @@
...
 	owner = null
 	..()
+
+/datum/species/dullahan/get_species_organs()
+	var/list/organs = ..()
+	organs[ORGAN_SLOT_EYES] = /obj/item/organ/eyes/dullahan
+	organs[ORGAN_SLOT_EARS] = /obj/item/organ/ears/dullahan
+	organs[ORGAN_SLOT_TONGUE] = /obj/item/organ/tongue/dullahan
+	organs[ORGAN_SLOT_BRAIN] = /obj/item/organ/brain/dullahan
+	return organs
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/human/species_types/dullahan.dm:145]: see above

File: code/modules/mob/living/carbon/human/species_types/ethereal.dm [hunk 10]
```diff
@@ -140,9 +138,15 @@
...
 	ethereal_charge = CLAMP(ethereal_charge + change, ETHEREAL_CHARGE_NONE, ETHEREAL_CHARGE_FULL)
 
 /datum/species/ethereal/proc/set_charge(var/change)
 	ethereal_charge = CLAMP(change, ETHEREAL_CHARGE_NONE, ETHEREAL_CHARGE_FULL)
+
+/datum/species/ethereal/get_species_organs()
+	var/list/organs = ..()
+	organs[ORGAN_SLOT_TONGUE] = /obj/item/organ/tongue/ethereal	
+	organs[ORGAN_SLOT_STOMACH] = /obj/item/organ/stomach/ethereal
+	return organs
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/human/species_types/ethereal.dm:152]: see above

File: code/modules/mob/living/carbon/human/species_types/felinid.dm [hunk 11]
```diff
@@ -65,33 +65,7 @@
...
-		if(new_species && new_species.mutanttail)
-			// Roundstart cat tail overrides new_species.mutanttail, reset it here.
-			new_species.mutanttail = initial(new_species.mutanttail)
-			if(new_species.mutanttail)
-				NT = new new_species.mutanttail
-		if(NT)
-			NT.Insert(H, drop_if_replaced = FALSE)
-		else
-			tail.Remove(H)
+//datum/species/human/felinid/on_species_loss(mob/living/carbon/H, datum/species/new_species, pref_load) - NOW OBSOLETE -
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/human/species_types/felinid.dm:68]: Commented out code

File: code/modules/mob/living/carbon/human/species_types/felinid.dm [hunk 12]
```diff
@@ -148,3 +122,9 @@
 
 	if(!silent)
 		to_chat(H, "You are no longer a cat.")
+
+/datum/species/human/felinid/get_species_organs()
+	var/list/organs = ..()
+	organs[ORGAN_SLOT_EARS] = mutantears
+	organs[ORGAN_SLOT_TAIL] = mutanttail
+	return organs
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/human/species_types/felinid.dm:130]: see above

File: code/modules/mob/living/carbon/human/species_types/flypeople.dm [hunk 13]
```diff
@@ -48,3 +45,10 @@
 	if(istype(weapon, /obj/item/melee/flyswatter))
 		return 29 //Flyswatters deal 30x damage to flypeople.
 	return 0
+
+/datum/species/fly/get_species_organs()
+	var/list/organs = ..()
+	organs[ORGAN_SLOT_TONGUE] = /obj/item/organ/tongue/fly	
+	organs[ORGAN_SLOT_STOMACH] = /obj/item/organ/stomach/fly
+	organs[ORGAN_SLOT_LIVER] = /obj/item/organ/liver/fly
+	return organs
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/human/species_types/flypeople.dm:54]: see above

File: code/modules/mob/living/carbon/human/species_types/golems.dm [hunk 14]
```diff
@@ -31,6 +30,11 @@
 	var/special_name_chance = 5
 	var/owner //dobby is a free golem
 
+/datum/species/golem/get_species_organs()
+	var/list/organs = ..()
+	organs[ORGAN_SLOT_ADAMANTINE_RESONATOR] = /obj/item/organ/adamantine_resonator
+	return organs
+
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/human/species_types/golems.dm:37]: see above

File: code/modules/mob/living/carbon/human/species_types/golems.dm [hunk 15]
```diff
@@ -65,12 +69,16 @@
...
 	fixed_mut_color = "4ed"
 	info_text = "As an <span class='danger'>Adamantine Golem</span>, you possess special vocal cords allowing you to \"resonate\" messages to all golems. Your unique mineral makeup makes you immune to most types of magic."
 	prefix = "Adamantine"
 	special_names = null
 
+/datum/species/golem/adamantine/get_species_organs()
+	var/list/organs = ..()
+	organs[ORGAN_SLOT_VOICE] = /obj/item/organ/vocal_cords/adamantine
+	return organs
+
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/human/species_types/golems.dm:81]: see above

File: code/modules/mob/living/carbon/human/species_types/golems.dm [hunk 16]
```diff
@@ -265,12 +273,16 @@
...
 	speedmod = 1 //faster
 	info_text = "As an <span class='danger'>Alloy Golem</span>, you are made of advanced alien materials: you are faster and regenerate over time. You are, however, only able to be heard by other alloy golems."
 	prefix = "Alien"
 	special_names = list("Outsider", "Technology", "Watcher", "Stranger") //ominous and unknown
 
+/datum/species/golem/alloy/get_species_organs()
+	var/list/organs = ..()
+	organs[ORGAN_SLOT_TONGUE] = /obj/item/organ/tongue/abductor
+	return organs
+
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/human/species_types/golems.dm:285]: see above

File: code/modules/mob/living/carbon/human/species_types/golems.dm [hunk 17]
```diff
@@ -828,10 +840,14 @@
...
 	special_step_sounds = list('sound/machines/clockcult/integration_cog_install.ogg', 'sound/magic/clockwork/fellowship_armory.ogg' )
-	mutantears = /obj/item/organ/ears/bronze
 	var/last_gong_time = 0
 	var/gong_cooldown = 150
 
+/datum/species/golem/bronze/get_species_organs()
+	var/list/organs = ..()
+	organs[ORGAN_SLOT_EARS] = /obj/item/organ/ears/bronze
+	return organs
+
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/human/species_types/golems.dm:850]: see above

File: code/modules/mob/living/carbon/human/species_types/golems.dm [hunk 18]
```diff
@@ -972,14 +988,19 @@
...
 	inherent_traits = list(TRAIT_RESISTHEAT,TRAIT_NOBREATH,TRAIT_RESISTCOLD,TRAIT_RESISTHIGHPRESSURE,TRAIT_RESISTLOWPRESSURE,TRAIT_NOFIRE,TRAIT_NOGUNS,TRAIT_RADIMMUNE,TRAIT_PIERCEIMMUNE,TRAIT_NODISMEMBER,TRAIT_FAKEDEATH)
 	species_language_holder = /datum/language_holder/golem/bone
 	info_text = "As a <span class='danger'>Bone Golem</span>, You have a powerful spell that lets you chill your enemies with fear, and milk heals you! Just make sure to watch our for bone-hurting juice."
 	var/datum/action/innate/bonechill/bonechill
 
+/datum/species/golem/bone/get_species_organs()
+	var/list/organs = ..()
+	organs[ORGAN_SLOT_TONGUE] = /obj/item/organ/tongue/bone
+	return organs
+
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/human/species_types/golems.dm:1003]: see above

File: code/modules/mob/living/carbon/human/species_types/jellypeople.dm [hunk 19]
```diff
@@ -19,6 +18,12 @@
 	inherent_factions = list("slime")
 	species_language_holder = /datum/language_holder/jelly
 
+/datum/species/jelly/get_species_organs()
+	var/list/organs = ..()
+	organs[ORGAN_SLOT_LUNGS] = /obj/item/organ/lungs/slime
+	return organs
+
+
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/human/species_types/jellypeople.dm:26]: see above

File: code/modules/mob/living/carbon/human/species_types/lizardpeople.dm [hunk 20]
```diff
@@ -24,6 +22,12 @@
 	inert_mutation = FIREBREATH
 	deathsound = 'sound/voice/lizard/deathsound.ogg'
 	species_language_holder = /datum/language_holder/lizard
+	
+/datum/species/lizard/get_species_organs()
+	var/list/organs = ..()
+	organs[ORGAN_SLOT_TONGUE] = /obj/item/organ/tongue/lizard
+	organs[ORGAN_SLOT_TAIL] = /obj/item/organ/tail/lizard
+	return organs
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/human/species_types/lizardpeople.dm:30]: see above

**Comment by eeSPee** [comment]
i'll return to this

**Pull request closed without merging**